### ISSUE TITLE
Inconsistent volume levels and sync with master volume

### ISSUE THREAD

**Issue by cdlenfert** [ejurgensen/forked-daapd#1011]
When I have a single speaker selected, the master volume would usually sync with that speaker. If I had multiple speakers selected, the master volume would sync with the loudest individual speaker, and adjust the other speakers proportionately. Lately however, I've notice that with a single speaker selected, the master doesn't sync with it, and often the individual speaker is basically muted because master volume is ~20% making the individual basically zero.
![image](https://user-images.githubusercontent.com/17277745/82228143-fa26c900-98e5-11ea-8063-42c54c0f3797.png)

I've noticed this somewhere in the last couple of updates, but that's not to say it hasn't been behaving this way longer.

... and I think I figured it out. After restarting my server I now see that the "Computer" speaker is selected and acting as the loudest speaker. However I have no local audio hardware hooked up to this device.
![image](https://user-images.githubusercontent.com/17277745/82228775-d1530380-98e6-11ea-89a4-ee52b126a284.png)

My Alsa config settings are as follows:
```
# ALSA device settings
# If you have multiple ALSA devices you can configure them individually via
# sections like the below. Make sure to set the "card name" correctly. See the
# README about ALSA for details. Note that these settings will override the ALSA
# settings in the "audio" section above.
#alsa "card name" {
        # Name - used in the speaker list in Remote
        # If not set, the card name will be used
#       nickname = "Computer"

        # Mixer channel to use for volume control
        # If not set, PCM will be used if available, otherwise Master
#       mixer = ""

        # Mixer device to use for volume control
        # If not set, the card name will be used
#       mixer_device = ""
#}
```
Previously I had my local audio settings all commented out. I notice in the default config that the `nickname = "Computer" is enabled.`. 
**My config**
```
# Name - used in the speaker list in Remote
#	nickname = "Pogo Local"
```

**Default** (not commented out)
```
# Name - used in the speaker list in Remote
nickname = "Computer"
```
I guess the change I was seeing was that before, the Computer "speaker" wasn't being auto-selected and now it looks like it is. I believe I discovered this behavior before and chalked it up to a fluke, disabled "Computer" and forgot about it. It seems to have come back on it's own however.

I believe the fix is setting the "Type of the output" to "disabled"
```
# Type of the output (alsa, pulseaudio, dummy or disabled)
	type = "disabled"
```

**Comment by ejurgensen** [comment]
What is your setting of "speaker_autoselect" in the config? If it is "no" or disabled it should not be possible for a speaker to be automatically selected (I do restart mine quite a lot). If your setting is no/disabled and you see autoselecting, is it then reproducible for you?

The same goes for the master volume issue - it should behave exactly like you say, i.e. match the loudest speaker. The first screenshot seems to indicate that not happening, but I need to know how to reproduce to fix this.

**Comment by ejurgensen** [comment]
Forget what I wrote about the master volume, I think I understand what happened, and what recent change caused this. Like you said it happens because a device that is not able to start was selected at the time you started playback. After you started playback it was detected as not working and set the speaker state to failed, which means you see it as not selected in the list. However, it was still used in the volume calculation, which is a bug. When you restarted, forked-daapd reset the speaker state, so the speaker again appeared as fully selected.

Still can't explain why it was selected in the first place, unless you have the setting I mentioned above enabled.

**Comment by cdlenfert** [comment]
@ejurgensen 
Thanks for the update. The volume behavior you describe is exactly what's going on.

My setting for autoselect was "yes" but was commented out. I believe that was the default until recently when the default was changed to "no"
```
#	speaker_autoselect = yes
```
I'm not sure how the "Computer" speaker was becoming selected either.

**Comment by cdlenfert** [comment]
@ejurgensen Now I remember how the Computer speaker is becoming selected. I have an iOS shortcut that starts a playlist in my kitchen, and to the Local audio device (default ID = 0). This worked at my offic where I only have my server connected via a local audio device (alsa + USB sound card), and at home where I had no local audio device (therefore only playing to the kitchen).  

The question that remains for me, is why (with the settings I had) was the "Computer" showing up as a speaker in the first place? I believe only the defaults should apply because everything is commented out. Is there some new-ish code that automatically tries to add local audio devices? Something like that sounds familiar in my memory.

**My relevant settings:**
```
# Local audio output
audio {
        # Name - used in the speaker list in Remote
#       nickname = "Pogo Local"

        # Type of the output (alsa, pulseaudio, dummy or disabled)
#       type = "alsa"

        # For pulseaudio output, an optional server hostname or IP can be
        # specified (e.g. "localhost"). If not set, connection is made via local
        # socket.
#       server = ""

        # Audio PCM device name for local audio output - ALSA only
#       card = "default"

        # Mixer channel to use for volume control - ALSA only
        # If not set, PCM will be used if available, otherwise Master.
#       mixer = ""

        # Mixer device to use for volume control - ALSA only
        # If not set, the value for "card" will be used.
#       mixer_device = ""

        # Enable or disable audio resampling to keep local audio in sync with
        # e.g. Airplay. This feature relies on accurate ALSA measurements of
        # delay, and some devices don't provide that. If that is the case you
        # are better off disabling the feature.
#       sync_disable = false

        # Here you can adjust when local audio is started relative to other
        # speakers, e.g. Airplay. Negative values correspond to moving local
        # audio ahead, positive correspond to delaying it. The unit is
        # milliseconds. The offset must be between -1000 and 1000 (+/- 1 sec).
#       offset_ms = 0

        # To calculate what and if resampling is required, local audio delay is
        # measured each second. After a period the collected measurements are
        # used to estimate drift and latency, which determines if corrections
        # are required. This setting sets the length of that period in seconds.
#       adjust_period_seconds = 100
}
```

```
# ALSA device settings
# If you have multiple ALSA devices you can configure them individually via
# sections like the below. Make sure to set the "card name" correctly. See the
# README about ALSA for details. Note that these settings will override the ALSA
# settings in the "audio" section above.
#alsa "card name" {
        # Name - used in the speaker list in Remote
        # If not set, the card name will be used
#       nickname = "Computer"

        # Mixer channel to use for volume control
        # If not set, PCM will be used if available, otherwise Master
#       mixer = ""

        # Mixer device to use for volume control
        # If not set, the card name will be used
#       mixer_device = ""
#}
```

**Comment by ejurgensen** [comment]
The default for forked-daapd has always been to show "Computer" for local audio, so that is why it shows up when everything is commented out. To remove "Computer" you have to set "type = disabled", like you did above.

**Comment by cdlenfert** [comment]
@ejurgensen: setting the `type = disabled` works for me. If I get rid of the local speaker ID in my JSON request I no longer get zero volume levels.

**Comment by ejurgensen** [comment]
This should be fixed now. The issue took me down a dangerous alley of "oh I should also fix/improve this", so it ended up being a major change. Hope it didn't include new bugs, but if you notice anything let me know.

**Comment by cdlenfert** [comment]
@ejurgensen I'm not sure if it's related, but I've reproduced an issue with volume levels using the JSON API on the latest version of FD 27.2. It may have be present in previous versions but I've been away from home for 2 months and haven't tested with my normal multi-speaker setup in that time.

### Here's what's happening: 

1. Select a speaker
2. Set the volume to 25% (or any value)
3. Use the JSON API `PUT /api/outputs/set` to select a **different** speaker (which disables the previously set speaker)
4. Volume jumps to 1 (as checked at `http://forked-daapd.local:3689/api/outputs`)

I'm noticing that all of the speakers other than the manually adjusted speaker from step 2 show a volume of 1. This behavior is not the same when using the `PUT /api/outputs/{id}/toggle` command where previously set volume levels are retained unless the master volume has become lower than the previous volume of the speaker (in which case the speaker volume is set to match the master and not to level 1).

My config is still set to disable the Computer/local output device via `type = disabled`, so no change there as far as I know.

**Comment by cdlenfert** [comment]
I also noticed that I had several CURL related package updates available on my server.

```root@pogoplug:~# apt list --upgradable
Listing... Done
curl/oldstable 7.52.1-5+deb9u11 armel [upgradable from: 7.52.1-5+deb9u10]
libcurl3/oldstable 7.52.1-5+deb9u11 armel [upgradable from: 7.52.1-5+deb9u10]
libcurl3-gnutls/oldstable 7.52.1-5+deb9u11 armel [upgradable from: 7.52.1-5+deb9u10]
libcurl4-gnutls-dev/oldstable 7.52.1-5+deb9u11 armel [upgradable from: 7.52.1-5+deb9u10]
```
Updating them didn't seem to make a difference however.

**Comment by ejurgensen** [comment]
Thanks for reporting. I’m away for a while but will take a look at it when I’m back.

**Comment by cdlenfert** [comment]
> Thanks for reporting. I’m away for a while but will take a look at it when I’m back.

Just let me know if you need anything when you get back to it. Does this issue need to be re-opened so it's on your radar? or would it help if I follow up?

**Comment by cdlenfert** [comment]
@ejurgensen this issue seems to have vanished as of this morning. I can't reproduce it. The only change I'm aware of since seeing it repeatedly for several days is that I restarted my Airport Extreme wireless router for the whole network. I also restarted a 2nd gen Airport Express independently, but I think it was the Extreme reboot that aligns most closely with the issue being resolved. I didn't update FD or restart the server it runs on. Very odd and surprising.
